go/simple/v2/ampli/ampli.go

```diff
@@ -0,0 +1,389 @@
+	return &EventWithAllProperties{
+		RequiredArray:   requiredArray,
+		RequiredBool:    requiredBool,
+		RequiredEnum:    requiredEnum,
+		RequiredInteger: requiredInteger,
+		RequiredNumber:  requiredNumber,
+		RequiredString:  requiredString,
+		optionals:       map[string]bool{},
+	}
+}
+
+func (stronglyTypedEvent *EventWithAllProperties) SetOptionalString(optionalString string) *EventWithAllProperties {
+	stronglyTypedEvent.OptionalString = optionalString
+	stronglyTypedEvent.optionals["OptionalString"] = true
+
+	return stronglyTypedEvent
+}
+
+func (stronglyTypedEvent *EventWithAllProperties) SetOptionalBool(optionalBool bool) *EventWithAllProperties {
+	stronglyTypedEvent.OptionalBool = optionalBool
+	stronglyTypedEvent.optionals["OptionalBool"] = true
+
+	return stronglyTypedEvent
+}
+
+func (stronglyTypedEvent EventWithAllProperties) ToEvent() amplitude.Event {
+	event := amplitude.Event{
+		EventType: "Event With All Properties",
+		EventProperties: map[string]interface{}{
+			"requiredArray":   stronglyTypedEvent.RequiredArray,
+			"requiredBool":    stronglyTypedEvent.RequiredBool,
+			"requiredEnum":    stronglyTypedEvent.RequiredEnum,
+			"requiredInteger": stronglyTypedEvent.RequiredInteger,
+			"requiredNumber":  stronglyTypedEvent.RequiredNumber,
+			"requiredString":  stronglyTypedEvent.RequiredString,
+		},
+	}
+
+	if _, present := stronglyTypedEvent.optionals["OptionalString"]; present {
+		event.EventProperties["optionalString"] = stronglyTypedEvent.OptionalString
+	}
+
+	if _, present := stronglyTypedEvent.optionals["OptionalBool"]; present {
+		event.EventProperties["optionalBool"] = stronglyTypedEvent.OptionalBool
+	}
+
+	return event
+}
+
+type Ampli struct {
+	Disabled bool
+	Client   amplitude.Client
+	mutex    sync.RWMutex
+}
+
+// Load initializes the Ampli wrapper.
+// Call once when your application starts.
+func (a *Ampli) Load(options LoadOptions) {
+	if a.Client != nil {
+		log.Default().Printf("Warn: Ampli is already initialized. Ampli.load() should be called once at application start up.")
+
+		return
+	}
+
+	if options.Client.isEmpty() {
+		options.Client = LoadClientOptions{
+			Configuration: DefaultConfiguration(),
+		}
+	}
+
+	var apiKey string
+	if options.Client.APIKey != "" {
+		apiKey = options.Client.APIKey
+	} else if options.Environment != "" {
+		apiKey = APIKey[options.Environment]
+	} else if options.Client.Configuration.APIKey != "" {
+		apiKey = options.Client.Configuration.APIKey
+	}
+
+	if !(apiKey != "" || options.Client.Instance != nil) {
+		log.Default().Printf("Error: Ampli.Load() requires option.Environment, " +
+			"and apiKey from either options.Instance.APIKey or APIKey[options.Environment], " +
+			"or options.Instance.Instance")
+	}
+
+	var configuration amplitude.Config
+	if !options.Client.Configuration.IsEmpty() {
```

Comment:
The logic is not simple. Did you test that all core config values (that a user do not set explicitly) will have default values (defined in core SDK)?

Reply:
Added tests for different load scenarios.

Only API key is required here, 5 ways to pass API key:
1. set by environment 
2. passed in `LoadClientOptions.APIKey`
3. passed in `LoadClientOptions.Client.Instance`
4. passed in `LoadClientOptions.Configuration`

`Amplitude.Client.Config.Logger` is an interface and cannot use `assert.Equal()`. To verify other default config values are set, it is equivalent to assert equal `FlushQueueSize` to default flush queue size, because all default config values are set together.